routes/cube_routes.js

```diff
@@ -1389,6 +1344,120 @@ async function bulkuploadCSV(req, res, cards, cube) {
+function compareCubes(req, res, cubeA, cubeB) {
+  if (!cubeA) {
+    req.flash('danger', 'Base cube not found');
+    res.status(404).render('misc/404', {});
+  } else if (!cubeB) {
+    req.flash('danger', 'Comparison cube was not found');
+    res.redirect('/cube/list/' + cubeA._id);
+  } else {
+    return new Promise(function(resolve, reject) {
+      let pids = [];
+      cubeA.cards.forEach(function(card, index) {
+        card.details = {
+          ...carddb.cardFromId(card.cardID),
+        };
+        if (!card.type_line) {
+          card.type_line = card.details.type;
+        }
+        if (card.details.tcgplayer_id && !pids.includes(card.details.tcgplayer_id)) {
+          pids.push(card.details.tcgplayer_id);
+        }
+      });
+      cubeB.cards.forEach(function(card, index) {
+        card.details = carddb.cardFromId(card.cardID);
+        if (!card.type_line) {
+          card.type_line = card.details.type;
+        }
+        if (card.details.tcgplayer_id && !pids.includes(card.details.tcgplayer_id)) {
+          pids.push(card.details.tcgplayer_id);
+        }
+      });
+      GetPrices(pids).then(function(price_dict) {
+        cubeA.cards.forEach(function(card, index) {
+          if (card.details.tcgplayer_id) {
+            if (price_dict[card.details.tcgplayer_id]) {
+              card.details.price = price_dict[card.details.tcgplayer_id];
+            }
+            if (price_dict[card.details.tcgplayer_id + '_foil']) {
+              card.details.price_foil = price_dict[card.details.tcgplayer_id + '_foil'];
+            }
+          }
+        });
+        cubeB.cards.forEach(function(card, index) {
+          if (card.details.tcgplayer_id) {
+            if (price_dict[card.details.tcgplayer_id]) {
+              card.details.price = price_dict[card.details.tcgplayer_id];
+            }
+            if (price_dict[card.details.tcgplayer_id + '_foil']) {
+              card.details.price_foil = price_dict[card.details.tcgplayer_id + '_foil'];
+            }
+          }
+        });
+        let in_both = [];
+        let only_a = cubeA.cards.slice(0);
+        let only_b = cubeB.cards.slice(0);
+        let a_names = only_a.map((card) => card.details.name);
+        let b_names = only_b.map((card) => card.details.name);
+
+        cubeA.cards.forEach(function(card, index) {
+          if (b_names.includes(card.details.name)) {
+            in_both.push(card);
+
+            only_a.splice(a_names.indexOf(card.details.name), 1);
+            only_b.splice(b_names.indexOf(card.details.name), 1);
+
+            a_names.splice(a_names.indexOf(card.details.name), 1);
+            b_names.splice(b_names.indexOf(card.details.name), 1);
+          }
+        });
+
+        let all_cards = in_both.concat(only_a).concat(only_b);
+        resolve({
+          in_both,
+          only_a,
+          only_b,
+          all_cards,
+        });
+      });
+    });
+  }
+}
+
+async function bulkReplaceCSV(req, res, list, cube) {
```

Comment:
helper functions are good - please keep them out of cube_routes. Move them into cubefn